src/sdx/pce/heuristic/heur.py

```diff
@@ -0,0 +1,274 @@
+import argparse
+import json
+
+# importing the module
+from datetime import datetime
+
+import numpy as np
+import prtpy
+
+from sdx.pce.load_balancing.te_solver import TESolver
+from sdx.pce.models import ConnectionRequest, TrafficMatrix
+from sdx.pce.utils.constants import Constants
+from sdx.pce.utils.random_connection_generator import RandomConnectionGenerator
+from sdx.pce.utils.random_topology_generator import RandomTopologyGenerator
+
+
+def random_graph(n, p, m):
+    graph_generator = RandomTopologyGenerator(n, p)
+    graph = graph_generator.generate_graph()
+
+    tm_generator = RandomConnectionGenerator(n)
+    tm = tm_generator.generate(m, 500, 2000, 50, 80).connection_requests
+    matrix = []
+    for rq in tm:
+        query = []
+        query.append(rq.source)
+        query.append(rq.destination)
+        query.append(rq.required_bandwidth)
+        query.append(rq.required_latency)
+        matrix.append(tuple(query))
+    return graph, matrix
+
+
+def matrix_to_connection(matrix):
+    traffic_matrix = TrafficMatrix(connection_requests=[])
+    for rq in matrix:
+        request = ConnectionRequest(
+            source=rq[0],
+            destination=rq[1],
+            required_bandwidth=rq[2],
+            required_latency=rq[3],
+        )
+        traffic_matrix.connection_requests.append(request)
+    return traffic_matrix
+
+
+class TEGroupSolver:
+    """ "
+    Class for a connection (TE matrix) splitting based heuristic
+    1. Sequential: binpacking like heuristic: first-fit(any-fit)-decreasing, refined first fit (4 classes): Online
+    2. Grouping: k-shortest-path based grouping: (k, then all other pairs sharing a same link) :
+        connection grouping: linear grouping (k largest items on); geometric grouping (interval [B/2^(r+1), B/2^2])
+    """
+
+    def __init__(self, topology, tm, cost, objective):
+        self.tm = tm
+        self.topology = topology
+        self.cost = cost
+        self.objective = objective
+        self.pad = 0
+
+    def connection_split(self, s, k):
+        s_tm = self.sort_tm()
+        if s == 0:
+            partition_tm = self.linear_partition(s_tm, k)
+        if s == 1:
+            partition_tm = self.geometry_partition(s_tm, k)
+        if s == 2:
+            partition_tm = self.kk_partition(s_tm, k)
+
+        return partition_tm
+
+    def sort_tm(self):
+        # sorted_tm = np.sort(np.asarray(self.tm), axis = -1)
+        print("tm shape:" + str(np.shape(self.tm)))
```

Comment:
f-strings are nicer:

```suggestion
        print(f"tm shape: {np.shape(self.tm)}")
```